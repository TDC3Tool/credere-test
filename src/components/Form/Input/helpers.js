import { findItemFromKey } from '../../../utils/array'

export const validator = (fn, el) => {
  const validation = fn(el.value)
  el.setCustomValidity(validation)

  if (!el.checkValidity()) {
    el.classList.add('border-danger')
  } else {
    el.classList.remove('border-danger')
  }

  return el
}

export const getFieldError = (fields = [], field) => {
  const el = findItemFromKey(fields, 'key', field)
  return el ? el.error : null
}

export const getFieldValue = (fields = [], field) => {
  const el = findItemFromKey(fields, 'key', field)
  return el ? el.value : ''
}

export const maxToday = () => new Date().toISOString().split('T')[0]
