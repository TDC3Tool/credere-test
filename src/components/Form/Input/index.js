import React, {
  useRef, useCallback, useContext, useEffect,
} from 'react'
import PropTypes from 'prop-types'
import Typography from '../../Typography'
import { validator, getFieldError } from './helper'
import formContext from '../context'
import {
  InputContainer, InputLabel, InputField, InputError,
} from './styles'

const Input = ({
  validation, field, label,
  className, ...props
}) => {
  const ref = useRef(null)
  const { fields, updateFields } = useContext(formContext)

  useEffect(() => {
    if (!ref.current.checkValidity()) {
      const { value } = ref.current
      updateFields({ key: field, value, error: null })
    }
  }, [ref, field, updateFields])

  const handleChange = useCallback(() => {
    ref.current = validator(validation, ref.current)
    const { validationMessage, value } = ref.current
    updateFields({ key: field, value, error: validationMessage })
  }, [validation, ref, field, updateFields])

  return (
    <InputContainer>
      {label && (
        <InputLabel htmlFor={field} className={className}>
          <Typography.Span>{label}</Typography.Span>
          {props.required ? ' *' : ''}
        </InputLabel>
      )}
      <InputField
        ref={ref}
        name={field}
        id={field}
        onChange={handleChange}
        {...props}
      />
      <InputError>
        <Typography.Span>{getFieldError(fields, field)}</Typography.Span>
      </InputError>
    </InputContainer>
  )
}

Input.propTypes = {
  required: PropTypes.bool,
  field: PropTypes.string.isRequired,
  label: PropTypes.string,
  className: PropTypes.string,
  validation: PropTypes.func,
}

Input.defaultProps = {
  validation: () => true,
  required: false,
  label: null,
  className: null,
}

export default Input
