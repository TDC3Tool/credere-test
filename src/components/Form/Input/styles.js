import styled from 'styled-components'
import {
  defaultRadius, defaultSpace, mobileSm, desktopSm, mobileLg,
} from '../../../globalStyles'

export const InputContainer = styled.div`
  margin-bottom: ${defaultSpace};
`

export const InputLabel = styled.label`
  display: block;
`
export const InputField = styled.input`
  display: inline-block;
  border: 1px solid #ccc;
  border-radius: ${defaultRadius};
  box-shadow: 0 3px 5px rgba(0,0,0,0.1);
  height: 40px;
  padding: 0 12px;
  color: #666;
  width: 95%;

  @media (min-width: ${mobileSm}) and (max-width: ${desktopSm}) {
    width: 94%;
  }

  @media screen and (max-width: ${mobileLg}) {
    width: 93%;
  }

  @media screen and (max-width: ${mobileSm}) {
    width: 90%;
  }
`
export const InputError = styled.div`
  display: block;
  color: #900;
  font-size: 0.8rem;
`
