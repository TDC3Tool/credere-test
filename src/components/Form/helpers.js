import { findItemFromKey } from '../../utils/array'

export const getFieldError = (fields = [], field) => {
  const el = findItemFromKey(fields, 'key', field)
  return el ? el.error : ''
}

export const getFieldValue = (fields = [], field) => {
  const el = findItemFromKey(fields, 'key', field)
  return el ? el.value : ''
}
