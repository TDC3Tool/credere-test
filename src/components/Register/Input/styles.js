import styled from 'styled-components'

export const InputContainer = styled.div`
  margin-bottom: 16px;
`

export const InputLabel = styled.label`
  display: block;
`
export const InputField = styled.input`
  display: inline-block;
  border: 1px solid #ccc;
  border-radius: 3px;
  box-shadow: 0 3px 5px rgba(0,0,0,0.1);
  height: 40px;
  padding: 0 12px;
  color: #666;
`
export const InputError = styled.div`
  display: block;
  color: #900;
  font-size: 0.8rem;
`
