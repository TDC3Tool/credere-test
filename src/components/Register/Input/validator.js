const validator = (fn, el) => {
  const validation = fn(el.value)
  el.setCustomValidity(validation)

  if (!el.checkValidity()) {
    el.classList.add('border-danger')
  } else {
    el.classList.remove('border-danger')
  }

  return el
}

export default validator
