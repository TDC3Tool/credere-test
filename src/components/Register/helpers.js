export const isUnderAge = (date) => {
  if (date) {
    const currentYear = new Date().getFullYear()
    const ageYear = new Date(date).getFullYear()

    return currentYear - ageYear < 18
  }

  return true
}

export const showCity = (state, driveLicense) => {
  if (state && driveLicense >= 0) {
    const sanitizedState = state.toLowerCase()
    const initialDriveLicense = driveLicense.slice(0, 1)

    if (sanitizedState === 'rn' && initialDriveLicense === '6') {
      return true
    }
  }

  return false
}
