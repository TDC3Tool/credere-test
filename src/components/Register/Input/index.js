import React, {
  useRef, useCallback, useState, useContext, useEffect,
} from 'react'
import PropTypes from 'prop-types'
import Typography from '../../Typography'
import validator from './validator'
import formContext from '../context'
import {
  InputContainer, InputLabel, InputField, InputError,
} from './styles'

const Input = ({
  validation, field, label,
  className, ...props
}) => {
  const ref = useRef(null)
  const [error, setError] = useState(null)
  const { updateFields } = useContext(formContext)

  useEffect(() => {
    if (!ref.current.checkValidity()) {
      const { validationMessage, value } = ref.current
      updateFields({ key: field, value, error: validationMessage })
    }
  }, [ref, field, updateFields])

  const handleChange = useCallback(() => {
    ref.current = validator(validation, ref.current)
    const { validationMessage, value } = ref.current
    setError(validationMessage)
    updateFields({ key: field, value, error: validationMessage })
  }, [validation, ref, field, updateFields])

  return (
    <InputContainer>
      {label && (
        <InputLabel htmlFor={field} className={className}>
          <Typography.Span>{label}</Typography.Span>
          {props.required ? ' *' : ''}
        </InputLabel>
      )}
      <InputField
        ref={ref}
        name={field}
        id={field}
        onChange={handleChange}
        {...props}
      />
      <InputError>
        <Typography.Span>{error}</Typography.Span>
      </InputError>
    </InputContainer>
  )
}

Input.propTypes = {
  required: PropTypes.bool,
  field: PropTypes.string.isRequired,
  label: PropTypes.string,
  className: PropTypes.string,
  validation: PropTypes.func,
}

Input.defaultProps = {
  validation: () => true,
  required: false,
  label: null,
  className: null,
}

export default Input
