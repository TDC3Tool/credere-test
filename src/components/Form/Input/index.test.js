import React from 'react'
import renderer from 'react-test-renderer'
import Input from '.'
import { maxToday } from './helpers'

describe('Input Component', () => {
  it('should render input component', () => {
    const element = renderer.create(
      <>
        <Input field="input" />
        <Input label="My Input" field="input" />
        <Input label="My Input" field="input" required />
        <Input className="my-class" label="My Input" field="input" required />
      </>,
    ).toJSON()
    expect(element).toMatchSnapshot()
  })
})

describe('Input Helpers', () => {
  it('should be today', () => {
    const today = new Date().toISOString().split('T')[0]
    expect(maxToday()).toBe(today)
  })

  it('should not be tomorrow', () => {
    let tomorrow = new Date()
    tomorrow.setDate(tomorrow.getDate() + 1)
    tomorrow = tomorrow.toISOString().split('T')[0]

    expect(maxToday()).not.toBe(tomorrow)
  })
})
