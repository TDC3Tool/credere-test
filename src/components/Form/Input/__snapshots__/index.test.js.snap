// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component should render input component 1`] = `
Array [
  <div
    className="border-sizing sc-ifAKCX jmfedv"
  >
    <input
      className="sc-bZQynM glljff"
      id="input"
      name="input"
      onChange={[Function]}
      required={false}
    />
    <div
      className="sc-gzVnrw bfjzjW"
    >
      <span
        className="sc-bxivhb bUZGRr"
      />
    </div>
  </div>,
  <div
    className="border-sizing sc-ifAKCX jmfedv"
  >
    <label
      className="sc-EHOje kyfnqM"
      htmlFor="input"
    >
      <span
        className="sc-bxivhb bUZGRr"
      >
        My Input
      </span>
      
    </label>
    <input
      className="sc-bZQynM glljff"
      id="input"
      name="input"
      onChange={[Function]}
      required={false}
    />
    <div
      className="sc-gzVnrw bfjzjW"
    >
      <span
        className="sc-bxivhb bUZGRr"
      />
    </div>
  </div>,
  <div
    className="border-sizing sc-ifAKCX jmfedv"
  >
    <label
      className="sc-EHOje kyfnqM"
      htmlFor="input"
    >
      <span
        className="sc-bxivhb bUZGRr"
      >
        My Input
      </span>
       *
    </label>
    <input
      className="sc-bZQynM glljff"
      id="input"
      name="input"
      onChange={[Function]}
      required={true}
    />
    <div
      className="sc-gzVnrw bfjzjW"
    >
      <span
        className="sc-bxivhb bUZGRr"
      />
    </div>
  </div>,
  <div
    className="my-class border-sizing sc-ifAKCX jmfedv"
  >
    <label
      className="sc-EHOje kyfnqM"
      htmlFor="input"
    >
      <span
        className="sc-bxivhb bUZGRr"
      >
        My Input
      </span>
       *
    </label>
    <input
      className="sc-bZQynM glljff"
      id="input"
      name="input"
      onChange={[Function]}
      required={true}
    />
    <div
      className="sc-gzVnrw bfjzjW"
    >
      <span
        className="sc-bxivhb bUZGRr"
      />
    </div>
  </div>,
]
`;
